# In this project i have done training on 20_newsgroup folder.
# I have done prediction on mini_newsgroup folder 

## Inbuilt naive bayes

In [0]:
import nltk #for stopwords
import numpy as np #numpy
from collections import Counter #to count frequency
from nltk.corpus import stopwords #for stopwords
stopwords=set(stopwords.words('english')) #all unique stopwords
All_words_dictionary={} #dictionary of all words and their frequency
wordsall=[] #to find all words after removing stopwords of each text file
wordsineachdoc=[] #to find all words in each individual class
words_after_stopwords=[] #to find all words after removing stopwords of each text file
wordswithoutspecialcharacters=[] #to find words without special ccharacters

#opening of all the text files in nested loop ,then doing folllwing steps :

import os
path1="20_newsgroups"
for l in os.listdir(path1) :
    path="20_newsgroups/"+l
    for filename in os.listdir(path):
        file1 = open("20_newsgroups/"+l+"/"+filename,"r")
        file=file1.read()
        words=file.split()                                  #splitting of all words
        for i in words:
            checkforspecialchar=i.isalpha()                 #checking for special characters
            if checkforspecialchar  == True :
                i=i.lower()                                 #converting words in lowercase after removing special chars
                wordswithoutspecialcharacters.append(i)
        for j in wordswithoutspecialcharacters :
            if j not in stopwords:
                words_after_stopwords.append(j)             # words after removing stopwords
                wordsall.append(j)
        wordswithoutspecialcharacters=[] 
    wordsineachdoc.append(wordsall)                         #collection of all words done classwise
    wordsall=[]


In [0]:
# counting all word's frequency and sorting them on descending order

All_words_dictionary_sorted=Counter(words_after_stopwords).most_common()

In [0]:
All_words_dictionary_sorted

[('apr', 19602),
 ('gmt', 17684),
 ('would', 15512),
 ('one', 12976),
 ('article', 12108),
 ('like', 9434),
 ('people', 8415),
 ('get', 8331),
 ('university', 8203),
 ('know', 7695),
 ('think', 7205),
 ('may', 7126),
 ('use', 6251),
 ('also', 6166),
 ('x', 6152),
 ('new', 6086),
 ('could', 5905),
 ('even', 5400),
 ('good', 5278),
 ('make', 4916),
 ('many', 4872),
 ('see', 4691),
 ('two', 4557),
 ('much', 4529),
 ('time', 4336),
 ('first', 4114),
 ('want', 4104),
 ('say', 4051),
 ('anyone', 3976),
 ('need', 3918),
 ('way', 3791),
 ('us', 3745),
 ('used', 3742),
 ('go', 3617),
 ('world', 3602),
 ('really', 3507),
 ('since', 3418),
 ('right', 3326),
 ('believe', 3308),
 ('still', 3290),
 ('going', 3216),
 ('something', 3190),
 ('computer', 3157),
 ('find', 3145),
 ('system', 3137),
 ('take', 3115),
 ('might', 3107),
 ('said', 3060),
 ('please', 3055),
 ('using', 3033),
 ('god', 2881),
 ('never', 2881),
 ('last', 2849),
 ('must', 2846),
 ('back', 2840),
 ('news', 2836),
 ('state', 2787),
 

In [0]:
#appending all data in the form of dictionary

final={}
final.update(All_words_dictionary_sorted)
final

{'apr': 19602,
 'gmt': 17684,
 'would': 15512,
 'one': 12976,
 'article': 12108,
 'like': 9434,
 'people': 8415,
 'get': 8331,
 'university': 8203,
 'know': 7695,
 'think': 7205,
 'may': 7126,
 'use': 6251,
 'also': 6166,
 'x': 6152,
 'new': 6086,
 'could': 5905,
 'even': 5400,
 'good': 5278,
 'make': 4916,
 'many': 4872,
 'see': 4691,
 'two': 4557,
 'much': 4529,
 'time': 4336,
 'first': 4114,
 'want': 4104,
 'say': 4051,
 'anyone': 3976,
 'need': 3918,
 'way': 3791,
 'us': 3745,
 'used': 3742,
 'go': 3617,
 'world': 3602,
 'really': 3507,
 'since': 3418,
 'right': 3326,
 'believe': 3308,
 'still': 3290,
 'going': 3216,
 'something': 3190,
 'computer': 3157,
 'find': 3145,
 'system': 3137,
 'take': 3115,
 'might': 3107,
 'said': 3060,
 'please': 3055,
 'using': 3033,
 'god': 2881,
 'never': 2881,
 'last': 2849,
 'must': 2846,
 'back': 2840,
 'news': 2836,
 'state': 2787,
 'without': 2749,
 'work': 2692,
 'well': 2669,
 'got': 2645,
 'someone': 2610,
 'better': 2545,
 'government': 253

In [0]:
#selecting top features
#here i have selected top 2000 features with their frequency

import itertools
selecteddict=dict(itertools.islice(final.items(),2000))
selectedfeatures=[]
for i in selecteddict:
    selectedfeatures.append(i)                      # appending top 2000 features into an array of features

selectedfeatures                           

['apr',
 'gmt',
 'would',
 'one',
 'article',
 'like',
 'people',
 'get',
 'university',
 'know',
 'think',
 'may',
 'use',
 'also',
 'x',
 'new',
 'could',
 'even',
 'good',
 'make',
 'many',
 'see',
 'two',
 'much',
 'time',
 'first',
 'want',
 'say',
 'anyone',
 'need',
 'way',
 'us',
 'used',
 'go',
 'world',
 'really',
 'since',
 'right',
 'believe',
 'still',
 'going',
 'something',
 'computer',
 'find',
 'system',
 'take',
 'might',
 'said',
 'please',
 'using',
 'god',
 'never',
 'last',
 'must',
 'back',
 'news',
 'state',
 'without',
 'work',
 'well',
 'got',
 'someone',
 'better',
 'government',
 'made',
 'problem',
 'another',
 'read',
 'usa',
 'information',
 'number',
 'sure',
 'come',
 'things',
 'part',
 'look',
 'point',
 'give',
 'little',
 'help',
 'every',
 'windows',
 'seems',
 'file',
 'data',
 'put',
 'question',
 'probably',
 'years',
 'different',
 'available',
 'space',
 'around',
 'long',
 'tell',
 'set',
 'least',
 'best',
 'program',
 'try',
 'anything',
 '

In [0]:
#finding how many times the feature comes in each class


data=np.empty([20,len(selectedfeatures)],dtype=int)
for i in range (20):
        for k in range (len(selectedfeatures)):
            data[i][k]=wordsineachdoc[i].count(selectedfeatures[k])
data  # here data is numpy array of shape 2000,20 where 2000( no of selected features) and 20 (no of class)

array([[1076,  893, 1018, ...,   10,    4,    3],
       [ 903,  873,  584, ...,    2,    9,    0],
       [ 786,  893,  419, ...,   10,    7,    2],
       ...,
       [ 855,  804, 1111, ...,   75,   12,   15],
       [1063,  866, 1384, ...,   11,    8,    5],
       [1059,  858, 1007, ...,    8,    8,   11]])

In [0]:
#numbering class/folder 1 as 1 and class/folder 20 as 20
#so for 20 classes we get array of 1 to 20

classes=[]
class_num=[]
path1="20_newsgroups"
for l in os.listdir(path1) :
    classes.append(l)
# class_num=np.empty([20,1],dtype=int)
for i in range (len(classes)):
    class_num.append(i+1)      #so for 20 classes we get  array of 1 to 20


classdict={}                    #dictionary of classes with their numbers
for i in range (len(classes)) :
    classdict[classes[i]]=class_num[i]

In [0]:
# this is to find for testing data (for mini_newsgroup folder )

wordsallt=[]
wordsineachtext=[]
words_after_stopwordst=[]
wordswithoutspecialcharacterst=[]
import os
path1t="mini_newsgroups"
for t in os.listdir(path1t) :
    patht="mini_newsgroups/"+t
    for filenamet in os.listdir(patht):
        file1 = open("mini_newsgroups/"+t+"/"+filenamet,"r")
        file=file1.read()
        wordst=file.split()                                     #splitting of all words
        for i in wordst:
            checkforspecialchart=i.isalpha()                    #checking for special characters
            if checkforspecialchart  == True :                  
                i=i.lower()                                     #converting words in lowercase after removing special chars
                wordswithoutspecialcharacterst.append(i)         
        for j in wordswithoutspecialcharacterst :         
            if j not in stopwords:
                wordsallt.append(j)                              # words after removing stopwords
        wordswithoutspecialcharacterst=[]
        wordsineachtext.append(wordsallt)                         #collection of all words in each text file 
        wordsallt=[]

In [0]:
# findind X_test
#here we will get numpy array of shape 2000,2000 where 2000 is (20*100 bcoz 100 text file in each 20 class) and 2000 is (no of selected features)

datat=np.empty([2000,len(selectedfeatures)])
for i in range (2000):
        for k in range (len(selectedfeatures)):
            datat[i][k]=wordsineachtext[i].count(selectedfeatures[k])
datat

array([[1., 1., 0., ..., 0., 0., 0.],
       [1., 1., 0., ..., 0., 0., 0.],
       [1., 1., 0., ..., 0., 0., 0.],
       ...,
       [2., 1., 0., ..., 0., 0., 0.],
       [1., 1., 0., ..., 0., 0., 0.],
       [1., 1., 1., ..., 1., 0., 0.]])

In [0]:
#running on inbuilt algorithm 

X_train=data
Y_train=class_num
X_test=datat
from sklearn.naive_bayes import MultinomialNB
clf=MultinomialNB()
clf.fit(X_train,Y_train)
Y_predinbuilt=clf.predict(X_test)

In [0]:
# making y _test as we know first 100 doc are in class 1 and so on .
# so y test is array having 2000 elements with first 100 elements as 1 and so on .so total we have 1-20(100 times each)
Y_test=[]

i=1
count=0
for i in range (1,21):
    while(count<100):
        Y_test.append(i)
        count=count+1
    count=0

In [0]:
clf.score(X_test,Y_test)


0.791

## For writing algo from scratch

In [0]:
# Fit function-It will make dictionary ,dictionary with top keys as classes and having nested dictionary with keys as 
#feature nos and value as frequency of that feature in respective classes.
                
def fit(X_train, Y_train):
    result = {}
    class_values = Y_train
    for current_class in class_values:
        result[current_class] = {}
        result["total_data"] = 19997
        num_features = X_train.shape[1]
        result[current_class]["total_count"] = data[current_class-1].sum()
        for j in range(num_features ):
            result[current_class][j] = {}
            #all_possible_values = set(X_train[:, j - 1])
            #for current_value in all_possible_values:
            result[current_class][j] = data[current_class-1][j]  #it is dictionary with top keys as classes and having
                                                                 #nested dictionary with keys as feature nos and value
                                                                 #as frequency of that feature in respective classes.
                
    return result 

In [0]:
# It is probability function which will calculate probability

def probability(dictionary, x, current_class):
    output=np.log(1/20)
    num_features = len(dictionary[current_class].keys()) - 1;
    for j in range(num_features ):
        if x[j]>0 :
            count_current_class_with_value_xj = dictionary[current_class][j] + 1
            count_current_class = dictionary[current_class]["total_count"] + num_features
            current_xj_probablity = np.log(count_current_class_with_value_xj)-np.log( count_current_class)
            output = output + current_xj_probablity
    return output

In [0]:
# For finding best classes

def predictSinglePoint(dictionary, x):
    class_values = dictionary.keys()
    best_p = -1e100
    best_class = -1
    first_run = True
    for current_class in class_values:
        if (current_class == "total_data"):
            continue
        p_current_class = probability(dictionary, x, current_class)
        if ( first_run or p_current_class > best_p):
            best_p = p_current_class
            best_class = current_class
        first_run = False
    return best_class

In [0]:
#Predict funcyion for predictions

def predict(dictionary, X_test):
    Y_predalgo = []
    for x in X_test:
        x_class = predictSinglePoint(dictionary, x)
        Y_predalgo.append(x_class)
    return Y_predalgo

In [0]:
dictionary = fit(X_train,Y_train)

In [0]:
Y_predalgoo = predict(dictionary,X_test)


# confusion matrix and report for inbulit Naivebayes

In [0]:
#confusion matrix and report for inbulit Naivebayes

from sklearn.metrics import classification_report, confusion_matrix

print(classification_report(Y_test,Y_predinbuilt))
inbuilt_naivebayes_matrix=confusion_matrix(Y_test,Y_predinbuilt)
print(inbuilt_naivebayes_matrix)

              precision    recall  f1-score   support

           1       0.66      0.80      0.72       100
           2       0.67      0.70      0.69       100
           3       0.71      0.79      0.75       100
           4       0.82      0.78      0.80       100
           5       0.80      0.85      0.83       100
           6       0.89      0.75      0.82       100
           7       0.74      0.88      0.80       100
           8       0.81      0.80      0.80       100
           9       0.81      0.92      0.86       100
          10       0.89      0.89      0.89       100
          11       0.97      0.90      0.93       100
          12       0.90      0.84      0.87       100
          13       0.73      0.77      0.75       100
          14       0.91      0.81      0.86       100
          15       0.89      0.86      0.87       100
          16       0.86      0.83      0.85       100
          17       0.63      0.83      0.72       100
          18       0.95    

In [0]:
# No of Correct predictions in inbuilt naive bayes
sum1=0
for i in range (20) :
    sum1+=inbuilt_naivebayes_matrix[i][i]
print(sum1)

1582


# confusion matrix and report for algorithm self implemented

In [0]:
#confusion matrix and report for algorithm self implemented

print(classification_report(Y_test,Y_predalgoo))
own_algo_matrix=confusion_matrix(Y_test,Y_predalgoo)
print(own_algo_matrix)

              precision    recall  f1-score   support

           1       0.64      0.83      0.72       100
           2       0.67      0.70      0.68       100
           3       0.73      0.83      0.78       100
           4       0.83      0.76      0.79       100
           5       0.75      0.82      0.78       100
           6       0.96      0.70      0.81       100
           7       0.74      0.84      0.79       100
           8       0.78      0.81      0.79       100
           9       0.79      0.90      0.84       100
          10       0.83      0.89      0.86       100
          11       0.97      0.88      0.92       100
          12       0.92      0.83      0.87       100
          13       0.70      0.77      0.73       100
          14       0.90      0.81      0.85       100
          15       0.89      0.87      0.88       100
          16       0.86      0.83      0.84       100
          17       0.66      0.85      0.75       100
          18       0.96    

In [0]:
# No of Correct predictions in iown implementation of naive bayes
sum2=0
for i in range (20) :
    sum2+=own_algo_matrix[i][i]
print(sum2)

1572


# Result (Comparison)

So we have done our project using both methods.From prediction reports and confusion matrix we can following :-
    1.for some classes inbult works better but for some algo works better.
    2.also our inbuilt naive bayes has done 1582 correct predictions out of 2000 
    3.Our own implemented algorithm has done 1572 correct predictions out of 2000
    4.Precision,recall,f1 score is almost same but slightly more for inbuilt naive bayes.
    
 CAUTION - The reports and results are based on selection of top 2000 features .After increasing selected features scores 
           will improve and will decrease if we decrease no of features.